amethyst_network/src/network_socket.rs

```diff
@@ -70,89 +45,90 @@ where
+        let server = Host::run(config.clone())?;
 
-        socket.set_nonblocking(true).map_err(|_| {
-            Error::new(
-                ErrorKind::Other,
-                "Unable to set `UdpSocket` to non-blocking mode",
-            )
-        })?;
+        let server_sender = NetSocketSystem::<E>::start_sending(&server);
+        let server_receiver = NetSocketSystem::<E>::start_receiving(server);
 
-        // this -> thread
-        let (tx1, rx1) = channel();
-        // thread -> this
-        let (tx2, rx2) = channel();
+        Ok(NetSocketSystem {
+            filters,
+            transport_sender: server_sender,
+            transport_receiver: server_receiver,
+            config,
+        })
+    }
 
-        thread::spawn(move || {
-            //rx1,tx2
-            let send_queue = rx1;
-            let receive_queue = tx2;
-            let mut socket = socket;
-
-            'outer: loop {
-                // send
-                for control_event in send_queue.try_iter() {
-                    match control_event {
-                        InternalSocketEvent::SendEvents { target, events } => {
-                            for ev in events {
-                                send_event(&ev, &target, &mut socket);
-                            }
+    /// Start a thread to send all queued packets.
+    fn start_sending(server: &Host) -> Sender<InternalSocketEvent<E>> {
+        let (tx, send_queue) = channel();
+
+        let udp_send_handler = server.udp_send_handle();
+
+        let send_channel = udp_send_handler.get_sender().clone();
+        thread::spawn(move || loop {
+            for control_event in send_queue.try_iter() {
+                match control_event {
+                    InternalSocketEvent::SendEvents { target, events } => {
+                        for ev in events {
+                            send_event(ev, target, &send_channel);
                         }
-                        InternalSocketEvent::Stop => break 'outer,
+                    }
+                    InternalSocketEvent::Stop => {
+                        break;
                     }
                 }
+            }
+        });
 
-                // receive
-                loop {
-                    match socket.recv() {
-                        // Data received
-                        Ok(Some(packet)) => {
-                            if let Err(_) = receive_queue.send(RawEvent::from(packet)) {
-                                error!("`NetworkSocketSystem` was dropped");
-                                break 'outer;
+        tx
+    }
+
+    /// Starts a thread which receives incoming packets and sends them onto the 'Receiver' channel.
+    fn start_receiving(server: Host) -> Receiver<Packet> {
+        let (receive_queue, rx) = channel();
+
+        thread::spawn(move || {
+            let receiver = server.udp_receive_handle();
+
+            let mut buffer = vec![ServerSocketEvent::Empty; receiver.max_throughput];
```

Comment:
Nit: I'd like to see this renamed to buffer_size not needed for this PR but just a note.